app/controllers/api/v1/questionnaires_controller.rb

```diff
@@ -0,0 +1,169 @@
+        @questionnaire.save
+        # Create node
+        tree_folder = TreeFolder.where(['name like ?', @questionnaire.display_type]).first
+        parent = FolderNode.find_by(node_object_id: tree_folder.id)
+        QuestionnaireNode.create(parent_id: parent.id, node_object_id: @questionnaire.id, type: 'QuestionnaireNode')
+        render json: @questionnaire, status: :created
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+     
+    end
+  end
+
+  # Remove a given questionnaire
+  # DELETE on /questionnaires/:id
+  def destroy
+    begin
+      @questionnaire = Questionnaire.find(params[:id])
+    rescue
+      render json: $ERROR_INFO
+    end
+    if @questionnaire
+      begin
+        name = @questionnaire.name
+        questions = @questionnaire.questions
+        questions.each do |question|
+          advices = question.question_advices
+          advices.each(&:delete)
+          question.delete
+        end
+        questionnaire_node = @questionnaire.questionnaire_node
+        questionnaire_node.delete if !questionnaire_node.nil?
+        @questionnaire.delete
+        render json: "The questionnaire \"#{name}\" has been successfully deleted."
+      rescue StandardError => e
+        render json: e.message
+      end
+    end
+  end
+
+  #Update a given questionnaire
+  # PUT on /questionnaires/:id
+  def update
+    # If 'Add' or 'Edit/View advice' is clicked, redirect appropriately
+      begin
+        # Save questionnaire information
+        @questionnaire = Questionnaire.find(params[:id])
+        @questionnaire.update_attributes(questionnaire_params)
+        # Save all questions
+        unless params[:question].nil?
+          params[:question].each_pair do |k, v|
+            @question = Question.find(k)
+            # example of 'v' value
+            # {"seq"=>"1.0", "txt"=>"WOW", "weight"=>"1", "size"=>"50,3", "max_label"=>"Strong agree", "min_label"=>"Not agree"}
+            v.each_pair do |key, value|
+              @question.send(key + '=', value) unless @question.send(key) == value
+            end
+            @question.save
+          end
+        end
+        render json: 'The questionnaire has been successfully updated!'
+      rescue StandardError
+        render json: $ERROR_INFO
+
+      end
+  end
+
+  # Create a clone of the given questionnaire, copying all associated
+  # questions. The name and creator are updated.
+  # POST on /questionnaires/copy/:id
+  def copy
+    # <Auth code add later>
+    # instructor_id = session[:user].instructor_id
+    instructor_id = 6
+    @questionnaire = Questionnaire.copy_questionnaire_details(params, instructor_id)
+    p_folder = TreeFolder.find_by(name: @questionnaire.display_type)
+    parent = FolderNode.find_by(node_object_id: p_folder.id)
+    QuestionnaireNode.find_or_create_by(parent_id: parent.id, node_object_id: @questionnaire.id)
```

Comment:
need to be more lean, similar to other endpoints

Reply:
Commit [f039d84](https://github.com/expertiza/reimplementation-back-end/pull/20/commits/f039d843a0176aa490bc542fe0cf6d42cca62d49) addresses this review